File: 1292-maximum-side-length-of-a-square-with-sum-less-than-or-equal-to-threshold/1292-maximum-side-length-of-a-square-with-sum-less-than-or-equal-to-threshold.go
func maxSideLength(mat [][]int, threshold int) int {
    m := len(mat)    
    n := len(mat[0])
    pref := make([][]int, m+1)
    for i := 0; i <= m; i++ {
        pref[i] = make([]int, n+1)
    }

    for i:= 1; i <= m; i++ {
        for j := 1; j <= n; j++ {
            pref[i][j] = mat[i-1][j-1] + pref[i-1][j] + pref[i][j-1] - pref[i-1][j-1]
        }
    }

    ans := 0
    lo, hi := 1, min(n, m)

    for lo <= hi {
        mid := (lo + hi) >> 1
        if check(pref, mid, threshold) {
            ans = mid
            lo = mid + 1
        } else {
            hi = mid - 1
        }
    }

    return ans
}

func check(pref [][]int, k, threshold int) bool {
    for i := k; i < len(pref); i++ {
        for j := k; j < len(pref[0]); j++ {
            sum := pref[i][j] - pref[i-k][j] - pref[i][j-k] + pref[i-k][j-k]
            if sum <= threshold {
                return true
            }
        }
    }
    return false
}

File: 1292-maximum-side-length-of-a-square-with-sum-less-than-or-equal-to-threshold/1292-maximum-side-length-of-a-square-with-sum-less-than-or-equal-to-threshold.rs
impl Solution {
    fn check(pref: &Vec<Vec<i32>>, k: usize, threshold: i32) -> bool {
        for i in k..pref.len() {
            for j in k..pref[0].len() {
                let sum = pref[i][j] - pref[i-k][j] - pref[i][j-k] + pref[i-k][j-k];
                if sum <= threshold {
                    return true
                }
            }
        }

        false
    }
    pub fn max_side_length(mat: Vec<Vec<i32>>, threshold: i32) -> i32 {
        let m = mat.len();
        let n = mat[0].len();

        let mut pref = vec![vec![0;n+1]; m+1];
        for i in 1..=m {
            for j in 1..=n {
                pref[i][j] = mat[i-1][j-1] + pref[i-1][j] + pref[i][j-1] - pref[i-1][j-1];
            }
        }

        let mut ans = 0;
        let mut lo = 1usize;
        let mut hi = m.min(n);
        while lo <= hi {
            let mid = (lo + hi) >> 1;
            if Self::check(&pref, mid, threshold) {
                ans = mid as i32;
                lo = mid+1;
            } else {
                hi = mid-1;
            }
        }
        ans
    }
}